packages/flutter/lib/src/services/text_input.dart

```diff
@@ -1319,61 +1376,82 @@ class TextInput {
     scheduleMicrotask(() {
       _hidePending = false;
       if (_currentConnection == null)
-        _channel.invokeMethod<void>('TextInput.hide');
+        _hide();
     });
   }
 
+  void _setClient(TextInputClient client, TextInputConfiguration configuration) {
+    for (final TextInputHandler handler in _inputHandlers)
```

Comment:
Thanks! Could you elaborate on this idea, please?

How would a `TextInputConnection` subclass control the processing order, or should the platform plugin have a say on it? Would this be an internal implementation detail, or do you mean bringing back parts of #69146 to let users implement custom `TextInputConnection` subclasses?

Should we also let the user decide whether a custom input handler should be processed before or after the platform input handler? Could it be something as simple as:

- `TextInput.addPreInputHandler(TextInputHandler handler)`
- `TextInput.addPostInputHandler(TextInputHandler handler)`

Or for more fine-grained control, perhaps associating each handler with a priority?

- `TextInput.addInputHandler(TextInputHandler handler, {int priority = kPlatformInputHandlerPriority})`

or
- `TextInput.addInputHandler(TextInputHandler handler, {TextInputHandlerPriority priority = TextInputHandlerPriority.platformInputPriority})`

Reply:
Yeah I meant something similar to #69146. The `TextInputConnection` subclass implemented by the in-app virtual keyboard can send the info to both the browser and the virtual keyboard implementation if `kIsWeb` is true.

If I understand this correctly `TextInputControl` is a `TextInputHandler` with additional capabilities and should be used to communicate with the "real" virtual keyboard. The reason for that is the web text input system requires info of the current input field even we're taking the user input from somewhere else.  
It seems UI-related communication needs to be in`TextInputControl`, I'm not sure that works. For example on web I think the web implementation still requires the info we sent via `setEditableSizeAndTransform` to position its input fields correctly.

Reply:
> Yeah I meant something similar to #69146. The TextInputConnection subclass implemented by the in-app virtual

An in-app virtual keyboard cannot reimplement `TextInputConnection` because that would break the connection to the platform plugin. The physical keyboard would no longer work as we learned in #69146.

> It seems UI-related communication needs to be in `TextInputControl`, I'm not sure that works. For example on web I think the web implementation still requires the info we sent via `setEditableSizeAndTransform` to position its input fields correctly.

Ok, thanks for pointing it out. Sounds like the separation between `TextInputHandler` and `TextInputControl` is not useful then. We could merge `TextInputControl` and `TextInputHandler` and deliver everything else except `show` and `hide` to both the custom and the platform input control.